File: backend/app/domain/activities/schemas/activity.py
"""Activity calendar and announcement payloads."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.domain.activities.models.activity import ActivityAudience, ActivityStatus


class ActivitySummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    summary: str | None = None
    location: str | None = None
    department_id: UUID
    department_slug: str | None = None
    department_name_zh: str | None = None
    department_accent_color: str | None = None
    starts_at: datetime
    ends_at: datetime | None = None
    status: ActivityStatus
    audience: ActivityAudience
    cover_image_url: str | None = None


class ActivityDetail(ActivitySummary):
    description: str | None = None
    capacity: int | None = None
    registration_url: str | None = None
    created_by_id: UUID | None = None
    created_by_name: str | None = None
    published_at: datetime | None = None
    announced_at: datetime | None = None


class ActivityCreate(BaseModel):
    title: str = Field(min_length=1, max_length=240)
    summary: str | None = Field(default=None, max_length=500)
    description: str | None = Field(default=None, max_length=8000)
    location: str | None = Field(default=None, max_length=240)
    department_id: UUID
    starts_at: datetime
    ends_at: datetime | None = None
    audience: ActivityAudience = ActivityAudience.DEPARTMENT
    capacity: int | None = Field(default=None, ge=1)
    registration_url: str | None = Field(default=None, max_length=500)
    cover_image_url: str | None = Field(default=None, max_length=500)


class ActivityUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=240)
    summary: str | None = Field(default=None, max_length=500)
    description: str | None = Field(default=None, max_length=8000)
    location: str | None = Field(default=None, max_length=240)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    audience: ActivityAudience | None = None
    status: ActivityStatus | None = None
    capacity: int | None = Field(default=None, ge=1)
    registration_url: str | None = Field(default=None, max_length=500)
    cover_image_url: str | None = Field(default=None, max_length=500)


class ActivityPublishRequest(BaseModel):
    """Publishing is separated from editing because it is what triggers member emails."""

    notify_by_email: bool = True


class AnnouncementSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    body: str
    department_id: UUID | None = None
    department_name_zh: str | None = None
    audience: ActivityAudience
    author_id: UUID | None = None
    author_name: str | None = None
    published_at: datetime | None = None
    is_pinned: bool
    emailed_at: datetime | None = None


class AnnouncementCreate(BaseModel):
    title: str = Field(min_length=1, max_length=240)
    body: str = Field(min_length=1, max_length=20000)
    department_id: UUID | None = None
    audience: ActivityAudience = ActivityAudience.DEPARTMENT
    is_pinned: bool = False
    send_email: bool = False
    publish_now: bool = True

File: frontend/src/shared/api/contracts/activities/activity.ts
import type {
  ActivityAudience,
  ActivityStatus,
} from '@/shared/api/contracts/activities/activityLifecycle'
import type { IsoDateTimeString, UuidString } from '@/shared/api/contracts/common/scalarTypes'

export interface ActivitySummary {
  id: UuidString
  title: string
  summary?: string | null
  location?: string | null
  department_id: UuidString
  department_slug?: string | null
  department_name_zh?: string | null
  department_accent_color?: string | null
  starts_at: IsoDateTimeString
  ends_at?: IsoDateTimeString | null
  status: ActivityStatus
  audience: ActivityAudience
  cover_image_url?: string | null
}

export interface ActivityDetail extends ActivitySummary {
  description?: string | null
  capacity?: number | null
  registration_url?: string | null
  created_by_id?: UuidString | null
  created_by_name?: string | null
  published_at?: IsoDateTimeString | null
  announced_at?: IsoDateTimeString | null
}

export interface ActivityCreate {
  title: string
  summary?: string | null
  description?: string | null
  location?: string | null
  department_id: UuidString
  starts_at: IsoDateTimeString
  ends_at?: IsoDateTimeString | null
  audience: ActivityAudience
  capacity?: number | null
  registration_url?: string | null
  cover_image_url?: string | null
}

export interface ActivityUpdate {
  title?: string | null
  summary?: string | null
  description?: string | null
  location?: string | null
  starts_at?: IsoDateTimeString | null
  ends_at?: IsoDateTimeString | null
  audience?: ActivityAudience | null
  status?: ActivityStatus | null
  capacity?: number | null
  registration_url?: string | null
  cover_image_url?: string | null
}

/** Publishing is separate from editing because it is what triggers member emails. */
export interface ActivityPublishRequest {
  notify_by_email: boolean
}

export interface ActivityQuery {
  department_id?: UuidString
  starts_after?: IsoDateTimeString
  starts_before?: IsoDateTimeString
  status?: ActivityStatus
}

export interface AnnouncementSummary {
  id: UuidString
  title: string
  body: string
  department_id?: UuidString | null
  department_name_zh?: string | null
  audience: ActivityAudience
  author_id?: UuidString | null
  author_name?: string | null
  published_at?: IsoDateTimeString | null
  is_pinned: boolean
  emailed_at?: IsoDateTimeString | null
}

export interface AnnouncementCreate {
  title: string
  body: string
  department_id?: UuidString | null
  audience: ActivityAudience
  is_pinned: boolean
  send_email: boolean
  publish_now: boolean
}
